src/main/java/io/confluent/idesidecar/restapi/messageviewer/RecordDeserializer.java

```diff
@@ -389,48 +420,77 @@ private static Optional<DecodedResult> maybeTrivialCase(
   ) {
     // If the byte array is null or empty, we return a NullNode or an empty string, respectively.
     if (bytes == null) {
-      return Optional.of(new DecodedResult(NullNode.getInstance(), null));
+      return Optional.of(new DecodedResult(
+          NullNode.getInstance(),
+          null,
+          null
+      ));
     }
     if (bytes.length == 0) {
-      return Optional.of(new DecodedResult(new TextNode(""), null));
+      return Optional.of(new DecodedResult(
+          new TextNode(""),
+          null,
+          null
+      ));
     }
 
     // If the first byte is not the magic byte, we try to parse the data as a JSON object
-    // or fall back to a string if parsing fails. Simple enough.
+    // or fall back to returning the raw bytes if parsing fails. Simple enough.
     if (bytes[0] != MAGIC_BYTE) {
-      return Optional.of(new DecodedResult(safeReadTree(bytes), null));
+      var wrappedJson = safeRead(bytes);
+      return Optional.of(
+          new DecodedResult(
+              wrappedJson.data(),
+              null,
+              new KeyOrValueMetadata(null, wrappedJson.dataFormat())
+          )
+      );
     }
 
     // If the first byte is the magic byte, but we weren't provided a schema registry client,
     // we can't decode the data, so we just return the raw bytes.
     if (schemaRegistryClient == null) {
+      var wrappedJson = safeRead(bytes);
       return Optional.of(new DecodedResult(
-          safeReadTree(bytes),
-          "The value references a schema but we can't find the schema registry"
+          wrappedJson.data(),
+          "The value references a schema but we can't find the schema registry",
+          new KeyOrValueMetadata(
+              null,
+              wrappedJson.dataFormat()
+          )
       ));
     }
 
     // Alright, let's move on to more complex cases.
     return Optional.empty();
   }
 
+  record WrappedJson(
+      JsonNode data,
+      DataFormat dataFormat
+  ) {
+  }
+
   /**
    * Try to read the byte array as a valid JSON value (object, array, string, number, boolean),
-   * falling back to a string if it fails.
+   * or encode it as a base64 string if it's not valid JSON (which means it's arbitrary bytes).
    * @param bytes The byte array to read
    * @return      A JsonNode representing the byte array as a JSON object, or a TextNode
    */
-  private static JsonNode safeReadTree(byte[] bytes) {
+  private static WrappedJson safeRead(byte[] bytes) {
     try {
-      return OBJECT_MAPPER.readTree(bytes);
+      return new WrappedJson(
+          OBJECT_MAPPER.readTree(bytes),
+          DataFormat.JSON
+      );
     } catch (IOException e) {
-      return TextNode.valueOf(new String(bytes, StandardCharsets.UTF_8));
+      return new WrappedJson(ByteArrayJsonUtil.asJsonNode(bytes), DataFormat.RAW_BYTES);
```

Comment:
Note: This is a major behavior change. We interpret bytes as base64 encoded string instead of UTF-8.